### PR TITLE
Vertical & Horizontal Fields for Unstructured

### LINKED ISSUE #1013: "Unstructured" k-field is not supported
There is currently no way to specify a k-field for grid type unstructured.
`UnstructuredFieldDimension` must be constructed with a neighbor chain of size >= 1, otherwise its API would be inconsistent (e.g. `UnstructuredFieldDimension::getDenseLocationType()`). Allowing an empty chain would require a change in the API with many consequences in passes, validators, codegen.
There is also no way to omit the `HorizontalFieldDimension` part in `FieldDimensions`. Many parts of the codebase, also, expect an horizontal field dimension to be present and they cast it to either cartesian or unstructured.
For reference, you can specify a k-field for cartesian with `CartesianFieldDimension` by setting the IJ mask to (0, 0).

It seems like there is no "quick" way to add support for this.

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

Introduces vertical (1D) & horizontal fields (2D) for unstructured situations. This required quite extensive changes throughout dawn.

* The horizontal field dimension in `sir::FieldDimensions` is now an optional.
* Vertical fields (1D) introduce a situation where we have a field with no location type. This affected all Validators, as well as the `PassLocalVarType`, which infers location types for local variables
* The unstructured codegens as well as de/serialization of `FieldDimensions` was adapted

## Testing

* New test covering the Validator Changes in `TestUnstructuredDimensionsCheckers`
* New test ensuring the correct working of the code generator in the unstructured integration tests, currently de-activated due to issue #1017
* New `dawn4py` test 

## Resolves / Enhances

Fixes #1013

### CODE REVIEW

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/src/dawn4py/serialization/utils.py [hunk 1]
```diff
@@ -240,6 +238,21 @@ def make_field(name: str, dimensions: FieldDimensions, is_temporary: bool = Fals
...
+    """
+    field = Field()
+    field.name = name
+    field.is_temporary = is_temporary
+    dims = FieldDimensions()
+    dims.mask_k = True
+    field.field_dimensions.CopyFrom(dims)
+    return field
+
+
```
[the remark this answers is not in the document]
Comment (Stagno) [reply dawn/src/dawn4py/serialization/utils.py:255]: So far the pattern that we used was calling `make_field` with a pre-constructed `FieldDimensions` object. However this is returned by a call to either `make_field_dimensions_cartesian` or `make_field_dimensions_unstructured`, none of which are really a good fit to specify the dimensions of a k-field.
I would add `make_field_dimensions_vertical` returning a `FieldDimensions` with horizontal dimension not specified and `mask_k=True`.
Then `make_field_dimensions_cartesian([0, 0 ,1])` could redirect to `make_field_dimensions_vertical()`.
I see no harm in keeping `make_vertical_field()` as syntactic sugar. To reuse code, it should just:
```
return make_field(name, make_field_dimensions_vertical(), is_temporary)
```

File: dawn/test/integration-test/dawn4py-tests/unstructured_masked_fields.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+OUTPUT_NAME = "unstructured_masked_fields"
+OUTPUT_FILE = f"{OUTPUT_NAME}.cpp"
+OUTPUT_PATH = f"{OUTPUT_NAME}.cpp"
+
+
+def main(args: argparse.Namespace):
+    interval = sir_utils.make_interval(
+        SIR.Interval.Start, SIR.Interval.End, 0, 0)
+
+    # create the out = in[i+1] statement
```
Comment (Stagno) [inline dawn/test/integration-test/dawn4py-tests/unstructured_masked_fields.py:40]:
```suggestion

```

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**